src/watch_filter.rs

```diff
@@ -0,0 +1,314 @@
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+    sync::{Arc, Mutex},
+    time::SystemTime,
+};
+
+use anyhow::{Context, Result};
+use glob::Pattern;
+use lazy_static::lazy_static;
+use watchexec::{
+    error::RuntimeError,
+    event::{
+        filekind::{FileEventKind, ModifyKind},
+        Event, Priority, Tag,
+    },
+    filter::Filterer,
+    signal::source::MainSignal::Interrupt,
+};
+
+lazy_static! {
+    /// A default set of ignore patterns that can be used by the consumer of WatchFilter.
+    pub static ref DEFAULT_IGNORE_PATTERNS: Vec<String> = vec![String::from("*.swp")];
+    /// Filter doesn't have a mutable reference so we need to store this state outside of filter.
+    static ref FILES_MODIFIED_AT: Mutex<HashMap<String, SystemTime>> = Mutex::new(HashMap::new());
+}
+
+#[derive(Debug)]
+pub struct WatchFilter {
+    path_patterns: Vec<Pattern>,
+    ignore_patterns: Vec<Pattern>,
+    process_start_time: SystemTime,
+    files_modified_at: Arc<Mutex<HashMap<String, SystemTime>>>,
+}
+
+impl WatchFilter {
+    pub fn new(
+        app_dir: PathBuf,
+        path_patterns: Vec<String>,
+        ignore_patterns: Vec<String>,
+    ) -> Result<Self> {
+        tracing::info!("watching relative to app dir: {app_dir:?}");
+        tracing::info!("watching path patterns: {path_patterns:?}");
+        tracing::info!("ignoring path patterns: {ignore_patterns:?}");
+
+        Ok(WatchFilter {
+            path_patterns: path_patterns
+                .iter()
+                .map(|path| WatchFilter::app_dir_pattern(&app_dir, path))
+                .collect::<Result<_>>()?,
+            ignore_patterns: ignore_patterns
+                .iter()
+                .map(|path| WatchFilter::app_dir_pattern(&app_dir, path))
+                .collect::<Result<_>>()?,
+            process_start_time: SystemTime::now(),
+            files_modified_at: Arc::new(Mutex::new(HashMap::new())),
+        })
+    }
+
+    fn app_dir_pattern(app_dir: &Path, path: &str) -> anyhow::Result<Pattern> {
+        let pat = app_dir.join(path);
+        let pat_str = pat
+            .to_str()
+            .with_context(|| format!("non-unicode pattern {path:?}"))?;
+        Pattern::new(pat_str).with_context(|| format!("invalid glob pattern {path:?}"))
+    }
+
+    fn has_valid_event_kind(&self, event: &Event) -> bool {
+        event.tags.iter().any(|tag| {
+            matches!(
+                tag,
+                Tag::FileEventKind(FileEventKind::Modify(ModifyKind::Data(_)))
```

Comment:
Could make this more concise, e.g.: `Tag::FileEventKind(FileEventKind::Modify(ModifyKind::Data(_) | ModifyKind::Name(_)) | FileEventKind::Create(_) | FileEventKind::Remove(_)))`